### ISSUE TITLE
[no-inferrable-types] suggests undefined is inferrable

### ISSUE THREAD

**Issue by RobinClowers** [typescript-eslint/typescript-eslint#544]
**Repro**

```JSON
{
  "rules": {
    "typescript/no-inferrable-types": "error"
  }
}
```

```typescript
const foo: undefined = undefined
```
**Expected Result**
Should not be an error, because typescript doesn't infer `undefined` in this case, it infers `any`.

**Actual Result**
The following error: "Type undefined trivially inferred from a undefined literal, remove type annotation  @typescript-eslint/no-inferrable-types"

**Additional Info**
I realize this is an edge case, we ran into it in test, where we wanted to be explicit about passing an undefined value for a certain parameter.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `22.0.0` |
| `TypeScript`                       | `3.3.0` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.11.3` |
| `yarn`                              | `1.15.2 |

**Comment by bradzacher** [comment]
might I ask why you're creating a constant variable with type undefined?

**Comment by csvn** [comment]
I believe `undefined` is only `any` when `strictNullChecks` is `false`. The rule would behave correctly if `strictNullChecks` is used (which I believe is recommended by Typescript). Don't know if the rule could so that it is not an error only with no strict null checks.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#new---strict-master-option

> Starting with TypeScript 2.3, the default tsconfig.json generated by tsc --init includes a "strict": true setting in the "compilerOptions" section. Thus, new projects started with tsc --init will by default have the highest level of type safety enabled.

**Comment by bradzacher** [comment]
Checking in the playground, you are correct - it only infers `undefined` as `any` if `strictNullChecks` is `false`.

We should have a flag to protect against this as we have access to the typescript config at lint time.

**Comment by bradzacher** [comment]
I'm going to close this for a few reasons:
- This rule does not use type information, so we don't have access to the tsconfig. This means the rule cannot know whether or not you have `strictNullChecks` on.
- It's an edge case that is hard to come by, as it's rare that someone creates a variable whose type is just `undefined` or `null`.
